conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/bitcoin/BitcoinIndexer.scala

```diff
@@ -0,0 +1,92 @@
+package tech.cryptonomic.conseil.indexer.bitcoin
+
+import java.util.concurrent.Executors
+
+import scala.concurrent.{ExecutionContext, Future}
+
+import cats.effect.{ContextShift, ExitCode, IO, Resource}
+import com.typesafe.scalalogging.LazyLogging
+import org.http4s.headers.Authorization
+import org.http4s.BasicCredentials
+import org.http4s.client.blaze.BlazeClientBuilder
+import slick.jdbc.PostgresProfile.api._
+import slickeffect.Transactor
+import slickeffect.transactor.config
+
+import tech.cryptonomic.conseil.common.util.DatabaseUtil
+import tech.cryptonomic.conseil.indexer.config.LorreConfiguration
+import tech.cryptonomic.conseil.common.config.Platforms
+import tech.cryptonomic.conseil.common.config.Platforms.BitcoinConfiguration
+import tech.cryptonomic.conseil.indexer.LorreIndexer
+import tech.cryptonomic.conseil.indexer.logging.LorreProgressLogging
+import tech.cryptonomic.conseil.common.rpc.RpcClient
+
+/**
+  * Class responsible for indexing data for Bitcoin Blockchain
+  *
+  * @param lorreConf Lorre configuration
+  * @param bitcoinConf Bitcoin configuration
+  */
+class BitcoinIndexer(
+    lorreConf: LorreConfiguration,
+    bitcoinConf: BitcoinConfiguration
+) extends LazyLogging
+    with LorreIndexer
+    with LorreProgressLogging {
+
+  private val executor = Executors.newFixedThreadPool(16)
+  implicit private val contextShift: ContextShift[IO] = IO.contextShift(ExecutionContext.fromExecutor(executor))
+  implicit val timer = IO.timer(ExecutionContext.global)
+  implicit private val httpEC: ExecutionContext = ExecutionContext.fromExecutor(executor) // Implicit is used to provide ExecutionContext for `stop`
+
+  /**
+    * Lorre for Bitcoin entry point. This method creates all the dependencies and wraps it into [[cats.Resource]].
+    */
+  def resource: Resource[IO, Unit] =
+    for {
+      httpClient <- BlazeClientBuilder[IO](httpEC).resource
+
+      rpcClient <- RpcClient.resource(
+        bitcoinConf.nodeConfig.url,
+        maxConcurrent = 8, // TODO: move to the configuration
+        httpClient, // TODO: wrap it into retry and logger middleware
+        Authorization(BasicCredentials(bitcoinConf.nodeConfig.username, bitcoinConf.nodeConfig.password))
+      )
+
+      tx <- Transactor
+        .fromDatabase[IO](IO.delay(DatabaseUtil.lorreDb))
+        .map(_.configure(config.transactionally)) // run operations in transaction
+
+      bitcoinOperations <- BitcoinOperations.resource(rpcClient, tx)
```

Comment:
Is this additional wrapping of every high-level component as a Resource needed?

Here the pattern looks like we want to create Resources for every possible thing and then create a sequence of those.
Does it makes sense for every operation to be treated as a resource, even if it's a processing result?
I feel it like we're forcing things a bit.

How would you see if we just wrap the resources, which needs proper allocation-deallocation, in some containing case class, and then use them in the `resource.use` block, to actually create processes that would produce IO values, as opposed to yet other IO resources?

Reply:
Please check it now :)